app/controllers/GalleryController.scala

```diff
@@ -0,0 +1,114 @@
+      case _ => Future.successful(
+        Redirect(s"/anonSignUp?url=/label/labelsByType?labelTypeId=" + labelTypeId + "&n=" + n + "&loadedLabels=" + loadedLabels)
+      )
+    }
+  }
+
+  def getLabelsBySeveritiesAndTags(labelTypeId: Int, n: Int, loadedLabels: String, severities: String, tags: String) = UserAwareAction.async { implicit request =>
+    request.identity match {
+      case Some(user) =>
+        val loadedLabelIds: Set[Int] = Json.parse(loadedLabels).as[JsArray].value.map(_.as[Int]).toSet
+        val severitiesToSelect: Set[Int] = Json.parse(severities).as[JsArray].value.map(_.as[Int]).toSet
+        val tagsToSelect: Set[String] = Json.parse(tags).as[JsArray].value.map(_.as[String]).toSet
+
+        //TODO: add a case for assorted labels
+        val labels: Seq[LabelValidationMetadata] = if (labelTypeId == 9) LabelTable.retrieveAssortedLabels(n, loadedLabelIds, Some(severitiesToSelect))
+                                                   else LabelTable.retrieveLabelsOfTypeBySeverityAndTags(labelTypeId, n, loadedLabelIds, severitiesToSelect, tagsToSelect)
+        val jsonList: Seq[JsObject] = labels.map(l => Json.obj(
+            "label" -> LabelTable.validationLabelMetadataToJson(l),
+            "imageUrl" -> getImageUrl(l.gsvPanoramaId, l.canvasWidth, l.canvasHeight, l.heading, l.pitch, l.zoom)
+          )
+        )
+
+        val labelList: JsObject = Json.obj("labelsOfType" -> jsonList)
+        Future.successful(Ok(labelList))
+      case _ => Future.successful(
+        Redirect(s"/anonSignUp?url=/label/labelsBySeveritiesAndTags?labelTypeId=" + labelTypeId + 
+                                                                    "&n=" + n + 
+                                                                    "&loadedLabels=" + loadedLabels +
+                                                                    "&severities=" + severities +
+                                                                    "&tags=" + tags)
+      )
+    }
+
+  }
+
+  /**
+   * TODO: try to refactor this somewhere else
+   *
+   * Retrieves the static image of the label panorama from the Google Street
+   * View Static API. Note that this returns the image of the panorama, but
+   * doesn't actually include the label. More information here:
+   * https://developers.google.com/maps/documentation/streetview/intro
+   *
+   * @param label Label to retrieve the static image of.
+   * @return  Image URL that represents the background of the label.
+   */
+  def getImageUrl(gsvPanoramaId: String, canvasWidth: Int, canvasHeight: Int, heading: Float, pitch: Float, zoom: Int): String = {
+    val url = "https://maps.googleapis.com/maps/api/streetview?" +
+      "pano=" + gsvPanoramaId +
+      "&size=" + canvasWidth + "x" + canvasHeight +
+      "&heading=" + heading +
+      "&pitch=" + pitch +
+      "&fov=" + getFov(zoom) +
+      "&key=" + VersionTable.getGoogleMapsAPIKey()
```

Comment:
You call this function for every label, and it actually does file I/O to get the API key, which has got to be slow. I should probably change this to being an environment variable, but for now maybe you could just [cache](https://www.playframework.com/documentation/2.3.x/ScalaCache#Accessing-the-Cache-API) the Google Maps API key? I link to the documentation here, but there are also a few examples throughout our code. You'll probably want to use the `Cache.getOrElse()` function. I would put the caching inside the `getGoogleMapsAPIKey()` function.

Reply:
Got it